keyboards/preonic/keymaps/brauner/keymap.c

```diff
@@ -0,0 +1,448 @@
+     */
+    [_COMPOSITOR] = LAYOUT_preonic_grid(
+      XXXXXXX, KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,   KC_8,    KC_9,    KC_0,    XXXXXXX,
+      XXXXXXX, KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,   KC_I,    KC_O,    KC_P,    XXXXXXX,
+      XXXXXXX, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,   KC_K,    KC_L,    KC_QUOT, KC_ENT,
+      XXXXXXX, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,   KC_COMM, KC_DOT,  KC_SLSH, XXXXXXX,
+      XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX, KC_SPC, XXXXXXX, XXXXXXX, XXXXXXX, XXXXXXX
+    )
+};
+
+static inline bool base_layer_hold_key(uint16_t keycode, keyrecord_t *record) {
+    if (!record->tap.count && record->event.pressed) {
+        tap_code16(keycode);
+        return false;
+    }
+    return true;
+}
+
+static inline bool toggle_layer(enum preonic_layers layer, keyrecord_t *record) {
+    if (record->event.pressed) {
+        layer_on(layer);
+    } else {
+        layer_off(layer);
+    }
+    return false;
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    switch (keycode) {
+        case QWERTY:
+            if (record->event.pressed) {
+                set_single_persistent_default_layer(_QWERTY);
+            }
+            return false;
+        case LOWER:
+	    return toggle_layer(_LOWER, record);
+        case RAISE:
+	    return toggle_layer(_RAISE, record);
+        case MANAGE:
+	    return toggle_layer(_MANAGE, record);
+        case COMPOSITOR:
+	    return toggle_layer(_COMPOSITOR, record);
+        case LT_CUT_X:
+	    return base_layer_hold_key(C(KC_X), record); /* cut on hold */
+        case LT_COPY_Y:
+	    return base_layer_hold_key(C(KC_C), record); /* copy on hold */
+        case LT_PASTE_P:
+	    return base_layer_hold_key(C(KC_V), record); /* paste on hold */
+        case LT_AE_A:
+	    return base_layer_hold_key(RALT(KC_Q), record); /* 'ä' on hold */
+        case LT_OE_O:
+	    return base_layer_hold_key(RALT(KC_P), record); /* 'ö' on hold */
+        case LT_UE_U:
+	    return base_layer_hold_key(RALT(KC_Y), record); /* 'ü' on hold */
+        case LT_SZ_S:
+	    return base_layer_hold_key(RALT(KC_S), record); /* 'ß' on hold */
+    }
+
+    return true;
+};
+
+bool     muse_mode      = false;
+uint8_t  last_muse_note = 0;
+uint16_t muse_counter   = 0;
+uint8_t  muse_offset    = 70;
+uint16_t muse_tempo     = 50;
+
+bool encoder_update_user(uint8_t index, bool clockwise) {
+    if (muse_mode) {
+        if (IS_LAYER_ON(_RAISE)) {
+            if (clockwise) {
+                muse_offset++;
+            } else {
+                muse_offset--;
+            }
+        } else {
+            if (clockwise) {
+                muse_tempo += 1;
+            } else {
+                muse_tempo -= 1;
+            }
+        }
+    } else {
+        if (clockwise) {
+            register_code(KC_PGDN);
+            unregister_code(KC_PGDN);
```

Comment:
you could use `tap_code` for these.  Simpler, and handles waits between, if enabled.

Reply:
done